### PR TITLE
Vectors for neighbors/edges for CycleGraphs

### PR DESCRIPTION
**Pull request by simonschoelly**
My other PR #10 was maybe a bit prematurely merged. On the other hand now I have my fourth PR for hacktoberfest, so this is also not too bad...

In this PR I replaced the neighbors and edge iterators with vectors `<: AbstractVector`.
One thing that I noticed was that I have now a lot of branching. Maybe the compiler is clever enough to convert them to some branchless expression or maybe that can be added in a later PR.

### CODE REVIEW

**Comment by simonschoelly** [comment]
Also uses `convert` instead of `SimpleGraph` for converting a `CycleGraph` to a `SimpleGraph`.

**Comment by matbesancon** [comment]
> My other PR #10 was maybe a bit prematurely merged. On the other hand now I have my fourth PR for hacktoberfest, so this is also not too bad...

Oops sorry, I thought this was a first good step and went with it. But also, you're welcome :stuck_out_tongue:

File: src/CycleGraph.jl [hunk 1]
```diff
@@ -76,95 +76,84 @@ function LG.has_edge(g::CycleGraph{T}, u, v) where {T}
...
 end
 
-function LG.iterate(iter::LG.SimpleGraphs.SimpleEdgeIter{G}, state) where {G <: CycleGraph}
+# TODO propagate inbounds
+function Base.getindex(edgevec::SimpleEdgeVector{V, G}, i::Int) where {V, G <: CycleGraph}
 
-    g = iter.g
-    nvg = nv(g)
-    T = eltype(g) 
+    i ∈ Base.OneTo(length(edgevec)) || throw(BoundsError(edgevec, i))
```
Comment (matbesancon) [inline src/CycleGraph.jl]:
```suggestion
    @boundscheck i ∈ Base.OneTo(length(edgevec)) || throw(BoundsError(edgevec, i))
```

Comment (matbesancon) [reply src/CycleGraph.jl]: that will do for the todo above

File: src/CycleGraph.jl [hunk 2]
```diff
@@ -76,95 +76,84 @@ function LG.has_edge(g::CycleGraph{T}, u, v) where {T}
...
-    T = eltype(g)
-
-    state == zero(T) && return nothing
-    return state, zero(T)
-end
+# TODO propagate inbounds
+function Base.getindex(nbs::OutNeighborVector{V, G}, i::Int) where {V, G <: CycleGraph}
 
-function Base.length(iter::OutNeighborsIter{V, <:CycleGraph}) where {V}
+    i ∈ Base.OneTo(length(nbs)) || throw(BoundsError(nbs, i))
```
Comment (matbesancon) [inline src/CycleGraph.jl]:
```suggestion
    @boundscheck i ∈ Base.OneTo(length(nbs)) || throw(BoundsError(nbs, i))
```

File: src/CycleGraph.jl [hunk 3]
```diff
@@ -76,95 +76,84 @@ function LG.has_edge(g::CycleGraph{T}, u, v) where {T}
...
+        v == nvg && return one(T)
+        return v - one(T)
+    end
+    # i == 2
+    v == one(T) && return nvg
+    v == nvg && return nvg - T(1)
+    return v + one(T)
 end
 
+Base.IndexStyle(::Type{<:OutNeighborVector{V, G}}) where {V, G <: CycleGraph} = IndexLinear()
```
Comment (matbesancon) [inline src/CycleGraph.jl:152]: is this required? Is `IndexLinear` not the default?

Comment (simonschoelly) [reply src/CycleGraph.jl:152]: No, it is `IndexCartesian`. Maybe I should put `IndexLinear` as default for these vectors though.

File: src/CycleGraph.jl (hunk 1 above)
Comment (simonschoelly) [reply src/CycleGraph.jl]: Has been implemented, also I added an `@inline` macro as `@boundcheck` seems to only work if the function is inlined.

File: src/CycleGraph.jl (hunk 2 above)
Comment (simonschoelly) [reply src/CycleGraph.jl]: see above

**Comment by simonschoelly** [comment]
I don't know why code coverage does not work for these two lines, maybe it is, because Julia realizes that their output is constant.

**Comment by simonschoelly** [comment]
Problems with coverage have been fixed now, turns out that calling `IndexStyle(edges)` is not enough, I also have to call `IndexStyle(typeof(edges))` even though the later always calls the former.

**Comment by matbesancon** [comment]
allright all things look good here

**Pull request merged**